doc/rst/developer/bestPractices/ContributorInfo.rst

```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+   wasted.
+
+#. If your work will require committing any third-party code that you are not
+   developing yourself (or code that you've developed as a standalone package),
+   alert the chapel-developers_ mailing list of this as, presently, such code
+   packages must be approved by Cray leadership before being committed.
+
+#. Sign a Chapel contributor's agreement and mail it, with your GitHub
+   ID, using the instructions here:
+
+   https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements/
 
 .. _Create new branch:
 
 Create new branch
-~~~~~~~~~~~~~~~~~
++++++++++++++++++
+
+* Develop your feature, bug fix, etc on your fork.  To create a new branch, use
+  the following command:
+
+.. code-block:: bash
+
+    git checkout -b <branch_name>
+
+  Using a concisely named branch is encouraged.
+
+.. _Develop and test contributions locally:
+
+Develop and test contributions locally
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Your contribution will take the form of a series of commits.  While including
+sensible commit messages is a good idea, it is more important to have a good
+merge message once the pull request is going in. Likewise, it is OK to have many
+small commits that reflect the history of development rather than commits for
+the feature.  See `Development commands`_ for how to perform some common
+operations during development.
+
+As you work, you will want to periodically bring in changes from the main Chapel
+project to your feature branch:
+
+.. code-block:: bash
+
+    git fetch upstream
+    git merge upstream/master
+
+    # or:
+    git pull upstream <branch_name>
+
+    # with feature branch checked out:
+    git merge [--no-ff] upstream/master
+
+If there are conflicts, you will be asked to resolve them. Once the affected
+files have been fixed, stage them with ``git add``, and then call ``git
+commit`` to finish the merge process.
+
+If you want to understand the changes that occurred upstream, see `Read commit
+messages`_ below.
+
+.. _Test your feature:
+
+Test your feature
++++++++++++++++++
+
+You will probably need to create new tests for your feature. See `Test System`_
+for more information on this process.  To do the most basic testing, you would
+do one of:
+
+.. code-block:: bash
+
+    # run Hello tests
+    make check
+
+    # run tests that end up in $CHPL_HOME/examples
+    start_test test/release/examples
+
+    # run all tests.  This will take a long time on a single desktop
+    start_test test/
+
+Any addition/change to the Chapel test system should pass testing when that
+test/directory is run with ``start_test`` (and performance tests should also
+pass testing for ``start_test -performance``).
+
+Contributors should be reasonably confident in the testing done on their code
+before asking for a review.  Should additional testing resources be needed, you
+can request help from a member of the core Chapel team when creating your pull
+request.
```

Comment:
Personally, I'd just focus the efforrt in this document on requesting that the reader clearly communicate to their reviewer the state of it - and point out that if it's doesn't work at least a little bit, getting feedback might not be that helpful. E.g. suggest that the PR message indicate major TODOs (like "check new tests work").
I'm happy with the idea of encouraging people to test to their satisfaction before making a PR, but I think it's a complicated issue enough that it might be misintepreted to exclude the way I use PRs (for example).